src/seccomp_notify.c

```diff
@@ -0,0 +1,225 @@
+		ctx->n_plugins++;
+
+	ctx->plugins = xmalloc0(sizeof(struct plugin) * (ctx->n_plugins + 1));
+
+	b = strdup(plugins);
+	if (b == NULL) {
+		pexit("Failed to strdup");
+		return -1;
+	}
+	for (s = 0, it = strtok_r(b, ":", &saveptr); it; s++, it = strtok_r(NULL, ":", &saveptr)) {
+		run_oci_seccomp_notify_plugin_version_cb version_cb;
+		run_oci_seccomp_notify_start_cb start_cb;
+		void *opq = NULL;
+
+		ctx->plugins[s].handle = dlopen(it, RTLD_NOW);
+		if (ctx->plugins[s].handle == NULL) {
+			pexitf("cannot load `%s`: %s", it, dlerror());
+			return -1;
+		}
+
+		version_cb = (run_oci_seccomp_notify_plugin_version_cb)dlsym(ctx->plugins[s].handle, "run_oci_seccomp_notify_version");
+		if (version_cb != NULL) {
+			int version;
+
+			version = version_cb();
+			if (version != 1) {
+				pexitf("invalid version supported by the plugin `%s`", it);
+				return -1;
+			}
+		}
+
+		ctx->plugins[s].handle_request_cb =
+			(run_oci_seccomp_notify_handle_request_cb)dlsym(ctx->plugins[s].handle, "run_oci_seccomp_notify_handle_request");
+		if (ctx->plugins[s].handle_request_cb == NULL) {
+			pexitf("plugin `%s` doesn't export `run_oci_seccomp_notify_handle_request`", it);
+			return -1;
+		}
+
+		start_cb = (run_oci_seccomp_notify_start_cb)dlsym(ctx->plugins[s].handle, "run_oci_seccomp_notify_start");
+		if (start_cb) {
+			int ret;
+
+			ret = start_cb(&opq, conf, sizeof(*conf));
+			if (ret != 0) {
+				pexitf("error loading `%s`", it);
+				return -1;
+			}
+		}
+		ctx->plugins[s].opaque = opq;
+	}
+
+	/* Change ownership.  */
+	*out = ctx;
+	ctx = NULL;
+	return 0;
+}
+
+int seccomp_notify_plugins_event(struct seccomp_notify_context_s *ctx, int seccomp_fd)
+{
+	size_t i;
+	int ret;
+
+	memset(ctx->sreq, 0, ctx->sizes.seccomp_notif);
+	memset(ctx->sresp, 0, ctx->sizes.seccomp_notif_resp);
+
+	ret = ioctl(seccomp_fd, SECCOMP_IOCTL_NOTIF_RECV, ctx->sreq);
+	if (ret < 0) {
+		nwarnf("Failed to read notification from %d", seccomp_fd);
+		return -1;
+	}
+
+	for (i = 0; i < ctx->n_plugins; i++) {
+		if (ctx->plugins[i].handle_request_cb) {
+			int handled = 0;
+			int ret;
+
+			ret = ctx->plugins[i].handle_request_cb(ctx->plugins[i].opaque, &ctx->sizes, ctx->sreq, ctx->sresp, seccomp_fd,
+								&handled);
+			if (ret != 0) {
+				nwarnf("Failed to handle seccomp notification from fd %d", seccomp_fd);
+				return -1;
+			}
+
+			switch (handled) {
+			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_NOT_HANDLED:
+				break;
+
+			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE:
+				goto send_resp;
+
+				/* The plugin will take care of it.  */
+			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_DELAYED_RESPONSE:
+				return 0;
+
+			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE_AND_CONTINUE:
+				ctx->sresp->flags |= SECCOMP_USER_NOTIF_FLAG_CONTINUE;
+				goto send_resp;
+
+			default:
+				pexitf("Unknown handler action specified %d", handled);
+				return -1;
+			}
+		}
+	}
+
+	/* No plugin could handle the request.  */
+	ctx->sresp->error = -ENOTSUP;
+	ctx->sresp->flags = 0;
```

Comment:
instead of the goto, can we add this in the handling of `RUN_OCI_SECCOMP_NOTIFY_HANDLE_NOT_HANDLED`
followed by breaks?

i.e.
```
			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_NOT_HANDLED:
			     /* No plugin could handle the request.  */
	                     ctx->sresp->error = -ENOTSUP;
	                    ctx->sresp->flags = 0;
                            break;
			case RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE:
                            break;

```

Reply:
we have to iterate all the plugins first.

I've replaced the goto statement